androidcomponents/androidcomponents.go

```diff
@@ -88,165 +97,165 @@ func Ensure(androidSdk *sdk.Model, gradlewPath string) error {
 	if err != nil {
 		return err
 	}
+	i := installer{
+		androidSdk,
+		sdkManager,
+		gradlewPath,
+	}
 
-	retryCount := 0
-	for true {
-		gradleCmd := command.New("./gradlew", "dependencies")
-		gradleCmd.SetStdin(strings.NewReader("y"))
-		gradleCmd.SetDir(filepath.Dir(gradlewPath))
+	return retry.Times(1).Wait(time.Second).Try(func(attempt uint) error {
+		if attempt > 0 {
+			log.Warnf("Retrying...")
+		}
+		return i.scanDependencies()
+	})
+}
 
-		log.Printf("Searching for missing SDK components using:")
-		log.Printf("$ %s", gradleCmd.PrintableCommandArgs())
+func (i installer) getDependencyCases() map[string]func(match string) error {
+	return map[string]func(match string) error{
+		`(Observed package id 'ndk-bundle' in inconsistent location)`: i.ndkInconsistentLocation,
+		`(NDK not configured)`:                                        i.ndkNotConfigured,
+		`failed to find target with hash string 'android-(.*)'\s*`:    i.target,
+		`failed to find Build Tools revision ([0-9.]*)\s*`:            i.buildTool,
+		`Could not find (com\.android\.support\..*)\.`:                i.extrasLib,
+	}
+}
 
-		if out, err := gradleCmd.RunAndReturnTrimmedCombinedOutput(); err != nil {
-			reader := strings.NewReader(out)
-			scanner := bufio.NewScanner(reader)
+func getDependenciesOutput(projectLocation string) (string, error) {
+	gradleCmd := command.New("./gradlew", "dependencies")
+	gradleCmd.SetStdin(strings.NewReader("y"))
+	gradleCmd.SetDir(projectLocation)
+	return gradleCmd.RunAndReturnTrimmedCombinedOutput()
+}
 
-			missingSDKComponentFound := false
+func (i installer) scanDependencies(foundMatches ...string) error {
+	out, err := getDependenciesOutput(filepath.Dir(i.gradlewPath))
+	if err == nil {
+		return nil
+	}
+	scanner := bufio.NewScanner(strings.NewReader(out))
+	for scanner.Scan() {
+		line := scanner.Text()
+		for pattern, callback := range i.getDependencyCases() {
+			re := regexp.MustCompile(pattern)
+			if matches := re.FindStringSubmatch(line); len(matches) == 2 {
+				if sliceutil.IsStringInSlice(matches[1], foundMatches) {
+					return fmt.Errorf("unable to solve a dependency installation for the output:\n%s", out)
+				}
+				if err := callback(matches[1]); err != nil {
+					fmt.Println(out)
+					return err
+				}
+				return i.scanDependencies(append(foundMatches, matches[1])...)
+			}
+		}
+	}
+	return nil
```

Comment:
In this case, the dependencies command's output does not match to any of our patterns, but the command failed. Should not we handle this situation?